### PR TITLE
[AArch64]Add convert and multiply-add SIMD&FP assembly/disassembly in…

### PR DESCRIPTION
**Pull request by CarolineConcatto**
…structions

This patch adds the following instructions:
Conversion between floating-point and integer:
  FCVT{AS, AU, MS, MU, NS, NU, PS, PU, ZS, ZU}
  {S,U}CVTF
Advanced SIMD three-register extension:
  FMMLA

According to https://developer.arm.com/documentation/ddi0602

Co-authored-by: Marian Lukac marian.lukac@arm.com
Co-authored-by: Spencer Abson spencer.abson@arm.com

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64InstrFormats.td [hunk 1]
```diff
@@ -5005,6 +5005,32 @@ multiclass FPToIntegerUnscaled<bits<2> rmode, bits<3> opcode, string asm,
...
+  // half-precision to 32-bit SIMD/FPR
+  def SHr :  BaseFPToIntegerUnscaled<0b11, rmode, opcode, FPR16, FPR32, asm,
+                                     []> {
+    let Inst{31} = 0; // 32-bit GPR flag
+  }
+
+  // half-precision to 64-bit SIMD/FPR
+  def DHr :  BaseFPToIntegerUnscaled<0b11, rmode, opcode, FPR16, FPR64, asm,
+                                     []> {
+    let Inst{31} = 1; // 32-bit GPR flag
```
Comment (jthackray) [inline llvm/lib/Target/AArch64/AArch64InstrFormats.td:5024]: Should this not say "64-bit GPR flag"?

File: llvm/lib/Target/AArch64/AArch64InstrFormats.td [hunk 2]
```diff
@@ -5005,6 +5005,32 @@ multiclass FPToIntegerUnscaled<bits<2> rmode, bits<3> opcode, string asm,
...
+  }
+
+  // single-precision to 64-bit SIMD/FPR
+  def DSr :  BaseFPToIntegerUnscaled<0b00, rmode, opcode, FPR32, FPR64, asm,
+                                     []> {
+    let Inst{31} = 1; // 32-bit GPR flag
```
Comment (jthackray) [inline llvm/lib/Target/AArch64/AArch64InstrFormats.td:5030]: Should this not say "64-bit GPR flag"?

File: llvm/lib/Target/AArch64/AArch64InstrFormats.td [hunk 3]
```diff
@@ -5190,6 +5218,32 @@ multiclass IntegerToFP<bit isUnsigned, string asm, SDPatternOperator node> {
...
+  // 32-bit to half-precision
+  def HSr: BaseIntegerToFPUnscaled<rmode, opcode, FPR32, FPR16, f16, asm, node> {
+    let Inst{31} = 0; // 32-bit GPR flag
+    let Inst{23-22} = 0b11; // 16-bit FPR flag
+  }
+
+  // 32-bit to double-precision
+  def DSr: BaseIntegerToFPUnscaled<rmode, opcode, FPR32, FPR64, f64, asm, node> {
+    let Inst{31} = 0; // 32-bit GPR flag
+    let Inst{23-22} = 0b01; // 16-bit FPR flag
```
Comment (jthackray) [inline llvm/lib/Target/AArch64/AArch64InstrFormats.td:5231]: The comment here ("16-bit FPR flag") and on line 5225 are the same, but the bit patterns different. Typo?

File: llvm/lib/Target/AArch64/AArch64InstrFormats.td (hunk 1 above)
Comment (CarolineConcatto) [reply llvm/lib/Target/AArch64/AArch64InstrFormats.td:5024]: Thank you for pointing that out Jonathan. In fact this is not even a GPR. I will update all these comments

File: llvm/lib/Target/AArch64/AArch64InstrFormats.td [hunk 4]
```diff
@@ -5419,6 +5447,32 @@ multiclass IntegerToFP<bit isUnsigned, string asm, SDPatternOperator node> {
...
+  // 64-bit to half-precision
+  def HDr: BaseIntegerToFPUnscaled<rmode, opcode, FPR64, FPR16, f16, asm, node> {
+    let Inst{31} = 1; // 64-bit FPR flag
+    let Inst{23-22} = 0b11; // 16-bit FPR flag
+  }
+
+  // 64-bit to single-precision
+  def SDr: BaseIntegerToFPUnscaled<rmode, opcode, FPR64, FPR32, f32, asm, node> {
+    let Inst{31} = 1; // 64-bit FPR flag
+    let Inst{23-22} = 0b00; // 16-bit FPR flag
```
Comment (jthackray) [inline llvm/lib/Target/AArch64/AArch64InstrFormats.td:5472]: This line says:
  `0b00; // 16-bit FPR flag`
but line 5466 above says
  `0b11; // 16-bit FPR flag`
Not sure if I'm seeing outdated diffs, or if this needs looking at?

Comment (CarolineConcatto) [reply llvm/lib/Target/AArch64/AArch64InstrFormats.td]: No, you were not! Thank you for pointing it out.

**Review by jthackray — APPROVED** [review]
LGTM

**Review by SpencerAbson — APPROVED** [review]
LGTM, thanks!

**Pull request merged**